plugins/cxaudio/fu-cxaudio-device.c

```diff
@@ -0,0 +1,881 @@
+					    G_USB_DEVICE_RECIPIENT_INTERFACE,
+					    HID_REPORT_SET,
+					    (HID_REPORT_TYPE_OUTPUT << 8) | report_number,
+					    FU_CXAUDIO_HID_INTERFACE,
+					    buf, bufsz, &actual_length,
+					    FU_CXAUDIO_USB_TIMEOUT, NULL, error)) {
+		return FALSE;
+	}
+	if (bufsz != actual_length) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "wrote 0x%x bytes of 0x%x",
+			     (guint) actual_length, bufsz);
+		return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_cxaudio_device_input_report (FuCxaudioDevice *self,
+				guint8 ReportID,
+				guint8 *buf,
+				guint16 bufsz,
+				GError **error)
+{
+	GUsbDevice *usb_device = fu_usb_device_get_dev (FU_USB_DEVICE (self));
+	gsize actual_length = 0;
+
+	g_return_val_if_fail (buf != NULL, FALSE);
+	g_return_val_if_fail (bufsz != 0, FALSE);
+
+	/* from device */
+	if (!g_usb_device_control_transfer (usb_device,
+					    G_USB_DEVICE_DIRECTION_DEVICE_TO_HOST,
+					    G_USB_DEVICE_REQUEST_TYPE_CLASS,
+					    G_USB_DEVICE_RECIPIENT_INTERFACE,
+					    HID_REPORT_GET,
+					    (HID_REPORT_TYPE_INPUT << 8) | ReportID,
+					    FU_CXAUDIO_HID_INTERFACE,
+					    buf, bufsz, &actual_length,
+					    FU_CXAUDIO_USB_TIMEOUT, NULL, error)) {
+		return FALSE;
+	}
+	if (g_getenv ("FWUPD_CXAUDIO_VERBOSE") != NULL)
+		fu_common_dump_raw (G_LOG_DOMAIN, "HID::READ", buf, bufsz);
+	if (bufsz != actual_length) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "read 0x%x bytes of expected 0x%x",
+			     (guint) actual_length, bufsz);
+		return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+typedef enum {
+	FU_CXAUDIO_OPERATION_READ,
+	FU_CXAUDIO_OPERATION_WRITE,
+	FU_CXAUDIO_OPERATION_LAST
+} FuCxaudioOperation;
+
+typedef enum {
+	FU_CXAUDIO_OPERATION_FLAG_NONE			= 0,
+	FU_CXAUDIO_OPERATION_FLAG_VERIFY		= (1 << 4),
+} FuCxaudioOperationFlags;
+
+static gboolean
+fu_cxaudio_device_operation (FuCxaudioDevice *self,
+			     FuCxaudioOperation operation,
+			     FuCxaudioMemKind mem_kind,
+			     guint32 addr,
+			     guint8 *buf,
+			     guint32 bufsz,
+			     FuCxaudioOperationFlags flags,
+			     GError **error)
+{
+	const guint32 idx_read = 0x1;
+	const guint32 idx_write = 0x5;
+	const guint32 payload_max = 0x20;
+	guint32 size = 0x02800;
+	g_autoptr(GPtrArray) chunks = NULL;
+
+	g_return_val_if_fail (bufsz > 0, FALSE);
+	g_return_val_if_fail (buf != NULL, FALSE);
+
+	/* check if memory operation is supported by device */
+	if (operation == FU_CXAUDIO_OPERATION_WRITE &&
+	    mem_kind == FU_CXAUDIO_MEM_KIND_CPX_ROM) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "trying to write unwritable section %u",
+			     mem_kind);
+		return FALSE;
+	}
+
+	/* check memory address - should be within valid range */
+	switch (mem_kind) {
+	case FU_CXAUDIO_MEM_KIND_EEPROM:
+		size = 0x20000;
+		break;
+	default:
+		break;
+	}
```

Comment:
Seems a bit heavy for a switch statement with only one case.

Reply:
Agreed, removed.